drivers/pwm/pwm_max31790.c

```diff
@@ -0,0 +1,685 @@
+	int result = i2c_reg_write_byte_dt(&config->i2c, address, value);
+
+	if (result != 0) {
+		LOG_ERR("unable to write register 0x%02X, error %i", address, result);
+	}
+
+	return result;
+}
+
+static bool max31790_convert_pwm_frequency_into_hz(uint16_t *result, uint8_t pwm_frequency)
+{
+	switch (pwm_frequency) {
+	case 0b0000:
+		*result = 25;
+		return true;
+	case 0b0001:
+		*result = 30;
+		return true;
+	case 0b0010:
+		*result = 35;
+		return true;
+	case 0b0011:
+		*result = 100;
+		return true;
+	case 0b0100:
+		*result = 125;
+		return true;
+	case 0b0101:
+		*result = 150; /* actually 149.7, according to the datasheet */
+		return true;
+	case 0b0110:
+		*result = 1250;
+		return true;
+	case 0b0111:
+		*result = 1470;
+		return true;
+	case 0b1000:
+		*result = 3570;
+		return true;
+	case 0b1001:
+		*result = 5000;
+		return true;
+	case 0b1010:
+		*result = 12500;
+		return true;
+	case 0b1011:
+		*result = 25000;
+		return true;
+	default:
+		LOG_ERR("invalid value %i for PWM frequency register", pwm_frequency);
+		return false;
+	}
+}
+
+static bool max31790_convert_pwm_frequency_into_register(uint8_t *result, uint32_t pwm_frequency)
+{
+	switch (pwm_frequency) {
+	case 25:
+		*result = 0b0000;
+		return true;
+	case 30:
+		*result = 0b0001;
+		return true;
+	case 35:
+		*result = 0b0010;
+		return true;
+	case 100:
+		*result = 0b0011;
+		return true;
+	case 125:
+		*result = 0b0100;
+		return true;
+	case 150: /* actually 149.7, according to the datasheet */
+		*result = 0b0101;
+		return true;
+	case 1250:
+		*result = 0b0110;
+		return true;
+	case 1470:
+		*result = 0b0111;
+		return true;
+	case 3570:
+		*result = 0b1000;
+		return true;
+	case 5000:
+		*result = 0b1001;
+		return true;
+	case 12500:
+		*result = 0b1010;
+		return true;
+	case 25000:
+		*result = 0b1011;
+		return true;
+	default:
+		LOG_ERR("invalid value %i for PWM frequency in Hz", pwm_frequency);
+		return false;
+	}
+}
+
+static int max31790_set_cycles_internal(const struct device *dev, uint32_t channel,
+					uint32_t period_count, uint32_t pulse_count,
+					pwm_flags_t flags)
+{
+	int result;
+	uint8_t pwm_frequency_channel;
+	enum max31790_register register_fan_configuration;
+	enum max31790_register register_pwm_target_duty_cycle_msb;
+	enum max31790_register register_pwm_target_duty_cycle_lsb;
+	enum max31790_register register_fan_dynamics;
+	uint8_t value_pwm_frequency;
+	uint8_t value_fan_configuration;
+	uint8_t value_fan_dynamics;
+	uint8_t value_speed_range = (flags >> PWM_MAX31790_FLAG_SPEED_RANGE_POS) &
+				    GENMASK(MAX37190_FANXDYNAMICS_SPEEDRANGE_LENGTH - 1, 0);
+	uint8_t value_pwm_rate_of_change =
+		(flags >> PWM_MAX31790_FLAG_PWM_RATE_OF_CHANGE_POS) &
+		GENMASK(MAX37190_FANXDYNAMICS_PWMRATEOFCHANGE_LENGTH - 1, 0);
+
+	if (!max31790_convert_pwm_frequency_into_register(&pwm_frequency_channel, period_count)) {
+		return -EINVAL;
+	}
+
+	result = max31790_read_register(dev, MAX37190_PWMFREQUENCY, &value_pwm_frequency);
+
```

Comment:
nit: could you drop these blank lines between function and return value check? I think that's how it's commonly done since the two are tightly tied in some way, makes the code a bit more compact and identifies the block as a coherent thing

Reply:
done